File: sample-conference-videochat/sample-conference-videochat/Components/DialogsScreen/DialogsViewController.m
//
//  DialogsViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "DialogsViewController.h"
#import "DialogsSelectionVC.h"
#import "CreateNewDialogViewController.h"
#import "UIViewController+InfoScreen.h"
#import "CacheManager.h"
#import "SVProgressHUD.h"
#import "Profile.h"
#import "MenuViewController.h"
#import "VideoSettingsViewController.h"
#import "AudioSettingsViewController.h"
#import "NSString+Chat.h"
#import "UIViewController+Alert.h"

@interface DialogsViewController () <ChatManagerDelegate, QBChatDelegate, ChatManagerConnectionDelegate>
//MARK: - Properties
@end

@implementation DialogsViewController

- (void)setupDialogs {
    self.chatManager.delegate = self;
    self.chatManager.connectionDelegate = self;
    [self.chatManager updateStorage];
    
    UILongPressGestureRecognizer *tapGestureDelete = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(tapEdit:)];
    tapGestureDelete.minimumPressDuration = 0.3;
    tapGestureDelete.delaysTouchesBegan = YES;
    [self.tableView addGestureRecognizer:tapGestureDelete];
}

- (void)setupNavigationBar {
    Profile *profile = [[Profile alloc] init];
    NSString *fullName = profile.fullName;
    NSUInteger ID = profile.ID;
    if (QBSession.currentSession.currentUser) {
        QBUUser *currentUser = QBSession.currentSession.currentUser;
        fullName = currentUser.fullName;
        ID = currentUser.ID;
    }
    
    NSCharacterSet *characterSet = [NSCharacterSet whitespaceCharacterSet];
    NSString *name = [fullName stringByTrimmingCharactersInSet:characterSet];
    NSString * firstLetter = name.firstLetter;
    
    UIButton *profileBarButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 28.0f, 28.0f)];
    profileBarButton.titleLabel.font = [UIFont systemFontOfSize:13.0f weight:UIFontWeightSemibold];
    [profileBarButton setTitle:firstLetter forState:UIControlStateNormal];
    [profileBarButton setTitle:firstLetter forState:UIControlStateHighlighted];
    profileBarButton.backgroundColor = [UIColor colorWithHexString:[NSString stringWithFormat:@"#%lX",
    (unsigned long)ID]];
    profileBarButton.layer.cornerRadius = 14.0f;
    [profileBarButton addTarget:self action:@selector(didTapMenu:) forControlEvents:UIControlEventTouchUpInside];

    UIBarButtonItem *menuItem = [[UIBarButtonItem alloc] initWithCustomView:profileBarButton];
    menuItem.tintColor = UIColor.whiteColor;
    self.navigationItem.leftBarButtonItem = menuItem;
    
    
    UIBarButtonItem *usersButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"add"]
                                                                        style:UIBarButtonItemStylePlain
                                                                       target:self
                                                                       action:@selector(didTapNewChat:)];
    self.navigationItem.rightBarButtonItem = usersButtonItem;
    usersButtonItem.tintColor = UIColor.whiteColor;
}

- (void)setupNavigationTitle {
    self.navigationItem.title = QBSession.currentSession.currentUser.fullName;
}

#pragma mark - Actions
- (void)didTapMenu:(UIBarButtonItem *)sender {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Chat" bundle:nil];
    MenuViewController *actionsMenuVC = [storyboard instantiateViewControllerWithIdentifier:@"MenuViewController"];
    actionsMenuVC.modalPresentationStyle = UIModalPresentationOverFullScreen;
    actionsMenuVC.menuType = TypeMenuAppMenu;

    __weak __typeof(self) weakSelf = self;
    
    QBUUser *currentUser = QBSession.currentSession.currentUser;
    
    MenuAction *userProfileAction = [[MenuAction alloc] initWithTitle:currentUser.name.capitalizedString action:ChatActionUserProfile handler:^(ChatAction action) {
        NSLog(@"User Profile");
    }];
    
    MenuAction *videoConfigAction = [[MenuAction alloc] initWithTitle:@"Video Configuration" action:ChatActionVideoConfig handler:^(ChatAction action) {
        __typeof(weakSelf)strongSelf = weakSelf;
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Settings" bundle:nil];
        VideoSettingsViewController *videoSettingsVC = [storyboard instantiateViewControllerWithIdentifier:@"VideoSettingsViewController"];
        [strongSelf.navigationController pushViewController:videoSettingsVC animated:YES];
    }];
    
    MenuAction *audioConfigAction = [[MenuAction alloc] initWithTitle:@"Audio Configuration" action:ChatActionAudioConfig handler:^(ChatAction action) {
        __typeof(weakSelf)strongSelf = weakSelf;
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Settings" bundle:nil];
        AudioSettingsViewController *audioSettingsVC = [storyboard instantiateViewControllerWithIdentifier:@"AudioSettingsViewController"];
        [strongSelf.navigationController pushViewController:audioSettingsVC animated:YES];
    }];
    
    MenuAction *appInfoAction = [[MenuAction alloc] initWithTitle:@"App Info" action:ChatActionAppInfo handler:^(ChatAction action) {
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf performSegueWithIdentifier:@"PresentInfoViewController" sender:nil];
    }];
    
    MenuAction *logoutAction = [[MenuAction alloc] initWithTitle:@"Logout" action:ChatActionLogout handler:^(ChatAction action) {
        __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf didTapLogout];
    }];

    [actionsMenuVC addAction:logoutAction];
    [actionsMenuVC addAction:appInfoAction];
    [actionsMenuVC addAction:audioConfigAction];
    [actionsMenuVC addAction:videoConfigAction];
    [actionsMenuVC addAction:userProfileAction];

    [self presentViewController:actionsMenuVC animated:NO completion:nil];
}

- (void)tapEdit:(UILongPressGestureRecognizer *)gestureReconizer {
    if (gestureReconizer.state == UIGestureRecognizerStateEnded) {
        return;
    }
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    DialogsSelectionVC *deleteVC = [storyboard instantiateViewControllerWithIdentifier:@"DialogsSelectionVC"];
    if (deleteVC) {
        deleteVC.action = ChatActionDelete;
        UINavigationController *navVC = [[UINavigationController alloc] initWithRootViewController:deleteVC];
        navVC.navigationBar.barTintColor = UIColor.mainColor;
        navVC.navigationBar.barStyle = UIBarStyleBlack;
        [navVC.navigationBar setTranslucent:NO];
        navVC.navigationBar.tintColor = UIColor.whiteColor;
        navVC.navigationBar.shadowImage = [UIImage imageNamed:@"navbar-shadow"];
        navVC.modalPresentationStyle = UIModalPresentationFullScreen;
        [self presentViewController:navVC animated:NO completion:^{
            [self.tableView removeGestureRecognizer:gestureReconizer];
        }];
    }
}

- (void)didTapNewChat:(UIButton *)sender {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    CreateNewDialogViewController *createNewDialogVC = [storyboard instantiateViewControllerWithIdentifier:@"CreateNewDialogViewController"];
    [self.navigationController pushViewController:createNewDialogVC animated:YES ];
}

#pragma mark Logout
- (void)didTapLogout {
    [SVProgressHUD showWithStatus:@"Logouting..."];
    __weak __typeof(self)weakSelf = self;
    [self deleteLastSubscriptionWithCompletion:^{
        [weakSelf.chatManager breakConnectionWithCompletion:^{
            [SVProgressHUD dismiss];
            [weakSelf.chatManager deactivateAutomaticMode];
            [Profile clear];
            [weakSelf.chatManager.storage clear];
            [CacheManager.instance clearCache];
            [NSUserDefaults.standardUserDefaults removeObjectForKey:kToken];
            if (self.onSignIn) {
                self.onSignIn();
            }
            [SVProgressHUD showSuccessWithStatus:@"Complited"];
        }];
    }];
}

- (void)deleteLastSubscriptionWithCompletion:(void(^)(void))completion {
    NSNumber *lastSubscriptionId = [NSUserDefaults.standardUserDefaults objectForKey:kSubscriptionID];
    if (lastSubscriptionId == nil) {
        if (completion) { completion(); }
        return;
    }
    
    [QBRequest deleteSubscriptionWithID:lastSubscriptionId.unsignedIntValue
                           successBlock:^(QBResponse * _Nonnull response) {
        [NSUserDefaults.standardUserDefaults removeObjectForKey:kSubscriptionID];
        Log(@"[%@] Delete Subscription request - Success",  NSStringFromClass(DialogsViewController.class));
        if (completion) { completion(); }
    } errorBlock:^(QBResponse * _Nonnull response) {
        Log(@"[%@] Delete Subscription request - Error",  NSStringFromClass(DialogsViewController.class));
        if (completion) { completion(); }
    }];
}

- (void)tableView:(UITableView *)tableView
         configureCell:(DialogCell *)cell
          forIndexPath:(NSIndexPath *)indexPath {
    tableView.allowsMultipleSelection = NO;
    cell.checkBoxImageView.hidden = YES;
    cell.checkBoxView.hidden = YES;
    cell.lastMessageDateLabel.hidden = NO;
    cell.contentView.backgroundColor = UIColor.clearColor;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    QBChatDialog *dialog = self.dialogs[indexPath.row];
    if (self.openChatScreen) {
        self.openChatScreen(dialog, NO);
    }
}

#pragma mark QBChatDelegate
- (void)chatDidReceiveMessage:(QBChatMessage *)message {
    [self.chatManager updateDialogWith:message.dialogID withMessage:message];
}

- (void)chatRoomDidReceiveMessage:(QBChatMessage *)message fromDialogID:(NSString *)dialogID {
    [self.chatManager updateDialogWith:dialogID withMessage:message];
}

- (void)chatDidReceiveSystemMessage:(QBChatMessage *)message {
    if ([self.chatManager.storage dialogWithID:message.dialogID]) {
        return;
    }
    [self.chatManager updateDialogWith:message.dialogID withMessage:message];
}

#pragma mark Chat Manager Delegate
- (void)chatManager:(ChatManager *)chatManager didUpdateStorage:(NSString *)message {
    [self reloadContent];
    [QBChat.instance addDelegate: self];
    [SVProgressHUD dismiss];
}

- (void)chatManager:(ChatManager *)chatManager didUpdateChatDialog:(QBChatDialog *)chatDialog {
    [self reloadContent];
    [SVProgressHUD dismiss];
}

- (void)chatManager:(ChatManager *)chatManager didFailUpdateStorage:(NSString *)message {
    [SVProgressHUD showErrorWithStatus:message];
}

- (void)chatManagerWillUpdateStorage:(ChatManager *)chatManager {
    if (self.navigationController.topViewController == self) {
        [SVProgressHUD show];
    }
}

//MARK: - ChatManagerConnectionDelegate
- (void)chatManagerConnect:(ChatManager *)chatManager {
    [SVProgressHUD showSuccessWithStatus:@"Connected"];
    [self.chatManager updateStorage];
}

- (void)chatManagerDisconnect:(ChatManager *)chatManager withLostNetwork:(BOOL)lostNetwork {
    if (lostNetwork == NO) { return; }
    if ([self.presentedViewController isKindOfClass:[Alert class]]) {
        Alert *alert = (Alert *)self.presentedViewController;
        if (alert.isPresented) {
            return;
        }
    }
    [self showAlertWithTitle:@"No Internet Connection" message:@"Make sure your device is connected to the internet" fromViewController:self];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/DialogsScreen/DialogsViewController.swift
//
//  DialogsViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 9/30/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
import Quickblox
import SVProgressHUD

class DialogsViewController: BaseDialogsViewController, DialogsView {
    //MARK: - Properties
    var onOpenChatScreenWithDialogID: ((String, Bool) -> Void)?
    var onSignIn: (() -> Void)?
    
    //MARK: - Setup
    override func setupNavigationBar() {
        let fullName = String(currentUser.fullName.capitalized.first ?? Character("U"))
        
        let profileBarButton = UIButton(frame: CGRect(x:0, y:0, width:28.0, height:28.0))
        profileBarButton.titleLabel?.font = .systemFont(ofSize: 13.0, weight: .semibold)
        profileBarButton.setTitle(fullName, for: .normal)
        profileBarButton.setTitle(fullName, for: .highlighted)
        profileBarButton.backgroundColor = currentUser.ID.generateColor()
        profileBarButton.layer.cornerRadius = 14.0
        profileBarButton.addTarget(self, action: #selector(didTapMenu(_:)), for: .touchUpInside)
        
        let leftBarButtonItem = UIBarButtonItem(customView: profileBarButton)
        self.navigationItem.leftBarButtonItem = leftBarButtonItem
        
        let usersButtonItem = UIBarButtonItem(image: UIImage(named: "add"),
                                              style: .plain,
                                              target: self,
                                              action: #selector(didTapNewChat(_:)))
        navigationItem.rightBarButtonItem = usersButtonItem
        usersButtonItem.tintColor = .white
    }
    
    override func setupNavigationTitle() {
        self.title = DialogsConstant.chats
    }
    
    override func setupDialogs() {
        chatManager.delegate = self
        SVProgressHUD.show()
        
        let tapGestureDelete = UILongPressGestureRecognizer(target: self, action: #selector(tapEdit(_:)))
        tapGestureDelete.minimumPressDuration = 0.3
        tapGestureDelete.delaysTouchesBegan = true
        tableView.addGestureRecognizer(tapGestureDelete)
        
        chatManager.updateStorage()
    }
    
    @objc func tapEdit(_ gestureReconizer: UILongPressGestureRecognizer) {
        guard gestureReconizer.state != UIGestureRecognizer.State.ended,
              let deleteVC = ScreenFactory().makeSelectionDialogsOutput() else { return }
        let navVC = UINavigationController(rootViewController: deleteVC)
        navVC.navigationBar.barTintColor = #colorLiteral(red: 0.2216441333, green: 0.4713830948, blue: 0.9869660735, alpha: 1)
        navVC.navigationBar.barStyle = .black
        navVC.navigationBar.shadowImage = UIImage(named: "navbar-shadow")
        navVC.navigationBar.isTranslucent = false
        navVC.modalPresentationStyle = .fullScreen
        present(navVC, animated: false) {
            self.tableView.removeGestureRecognizer(gestureReconizer)
        }
    }
    
    @objc func didTapNewChat(_ sender: UIBarButtonItem) {
        guard let createNewDialogVC = ScreenFactory().makeCreateNewDialogOutput() else { return }
        navigationController?.pushViewController(createNewDialogVC, animated: true)
    }
    
    @objc func didTapMenu(_ sender: UIBarButtonItem) {
        guard let actionsMenuVC = ScreenFactory().makeActionsMenuOutput() else { return }
        actionsMenuVC.typeActionsMenuVC = .appMenu
        actionsMenuVC.modalPresentationStyle = .overFullScreen
        
        let userProfileAction = MenuAction(title: Profile().fullName.capitalized, action: .userProfile) { (action) in
            print("User Profile")
        }
        let videoConfigAction = MenuAction(title: "Video Configuration", action: .videoConfig) { [weak self] (action) in
            guard let videoSettingVC = ScreenFactory().makeVideoSettingsOutput() else { return }
            self?.navigationController?.pushViewController(videoSettingVC, animated: true)
        }
        let audioConfigAction = MenuAction(title: "Audio Configuration", action: .audioConfig) { [weak self] (action) in
            guard let audioSettingVC = ScreenFactory().makeAudioSettingsOutput() else { return }
            self?.navigationController?.pushViewController(audioSettingVC, animated: true)
        }
        let appInfoAction = MenuAction(title: "App Info", action: .appInfo) { [weak self] (action) in
            self?.performSegue(withIdentifier: DialogsConstant.infoSegue, sender: nil)
        }
        let logoutAction = MenuAction(title: "Logout", action: .logout) { [weak self] (action) in
            self?.didTapLogout()
        }
        
        actionsMenuVC.addAction(userProfileAction)
        actionsMenuVC.addAction(videoConfigAction)
        actionsMenuVC.addAction(audioConfigAction)
        actionsMenuVC.addAction(appInfoAction)
        actionsMenuVC.addAction(logoutAction)
        
        present(actionsMenuVC, animated: false)
    }
    
    private func didTapLogout() {
        SVProgressHUD.show(withStatus: "Logouting...")
        SVProgressHUD.setDefaultMaskType(.clear)
        
        guard let uuidString = UIDevice.current.identifierForVendor?.uuidString else {
            return
        }
        #if targetEnvironment(simulator)
        disconnectUser()
        #else
        QBRequest.subscriptions(successBlock: { [weak self] (response, subscriptions) in
            guard let subscriptions = subscriptions, subscriptions.isEmpty == false else {
                self?.disconnectUser()
                return
            }
            let deleteSubscriptionsGroup = DispatchGroup()
            for subscription in subscriptions {
                if let subscriptionsUIUD = subscription.deviceUDID,
                   subscriptionsUIUD == uuidString {
                    deleteSubscriptionsGroup.enter()
                    QBRequest.deleteSubscription(withID: subscription.id) { (response) in
                        deleteSubscriptionsGroup.leave()
                        debugPrint("[NotificationsProvider] Unregister Subscription request - Success")
                    } errorBlock: { (response) in
                        deleteSubscriptionsGroup.leave()
                        debugPrint("[NotificationsProvider] Unregister Subscription request - Error")
                    }
                }
            }
            deleteSubscriptionsGroup.notify(queue: DispatchQueue.main) {
                self?.disconnectUser()
            }
        }) { response in
            if response.status.rawValue == 404 {
                self.showLoginScreen()
            } else if let error = response.error?.error {
                SVProgressHUD.showError(withStatus: error.localizedDescription)
            }
        }
        #endif
    }
    
    //MARK: - Internal Methods
    //MARK: - logOut flow
    fileprivate func showLoginScreen() {
        Profile.clear()
        chatManager.storage.clear()
        CacheManager.shared.clearCache()
        onSignIn?()
        SVProgressHUD.showSuccess(withStatus: "Complited")
    }
    
    private func disconnectUser() {
        chatManager.disconnect()
        logOut()
    }
    
    private func logOut() {
        QBRequest.logOut(successBlock: { [weak self] response in
            self?.showLoginScreen()
        }) { response in
            debugPrint("[DialogsViewController] logOut error: \(response)")
        }
    }
    
    // MARK: - UITableViewDataSource
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dialogs.count
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 60.0
    }
    
    override func tableView(_ tableView: UITableView,
                            configureCell cell: DialogCell,
                            for indexPath: IndexPath) {
        
        tableView.allowsMultipleSelection = false
        cell.checkBoxImageView.isHidden = true
        cell.checkBoxView.isHidden = true
        cell.lastMessageDateLabel.isHidden = false
        cell.contentView.backgroundColor = .clear
        cell.lastMessageDateLabel.isHidden = false
    }
    
    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let dialog = dialogs[indexPath.row]
        if let dialogID = dialog.id {
            onOpenChatScreenWithDialogID?(dialogID, false)
        }
    }
}

// MARK: - QBChatDelegate
extension DialogsViewController: QBChatDelegate {
    func chatRoomDidReceive(_ message: QBChatMessage, fromDialogID dialogID: String) {
        chatManager.updateDialog(with: dialogID, with: message)
    }
    
    func chatDidReceive(_ message: QBChatMessage) {
        guard let dialogID = message.dialogID else {
            return
        }
        chatManager.updateDialog(with: dialogID, with: message)
    }
    
    func chatDidReceiveSystemMessage(_ message: QBChatMessage) {
        guard let dialogID = message.dialogID,
              chatManager.storage.dialog(withID: dialogID) == nil else {
            return
        }
        chatManager.updateDialog(with: dialogID, with: message)
    }
    
    func chatServiceChatDidFail(withStreamError error: Error) {
        SVProgressHUD.showError(withStatus: error.localizedDescription)
    }
    
    func chatDidAccidentallyDisconnect() {
    }
    
    func chatDidNotConnectWithError(_ error: Error) {
        SVProgressHUD.showError(withStatus: error.localizedDescription)
    }
    
    func chatDidDisconnectWithError(_ error: Error?) {
        if let error = error {
            SVProgressHUD.showError(withStatus: error.localizedDescription)
        }
    }
    
    func chatDidConnect() {
        if QBChat.instance.isConnected == true {
            chatManager.updateStorage()
            SVProgressHUD.showSuccess(withStatus: "SA_STR_CONNECTED".localized)
        }
    }
    
    func chatDidReconnect() {
        SVProgressHUD.show(withStatus: "SA_STR_CONNECTED".localized)
        if QBChat.instance.isConnected == true {
            chatManager.updateStorage()
            SVProgressHUD.showSuccess(withStatus: "SA_STR_CONNECTED".localized)
        }
    }
}

// MARK: - ChatManagerDelegate
extension DialogsViewController: ChatManagerDelegate {
    func chatManager(_ chatManager: ChatManager, didUpdateChatDialog chatDialog: QBChatDialog) {
        reloadContent()
        SVProgressHUD.dismiss()
    }
    
    func chatManager(_ chatManager: ChatManager, didFailUpdateStorage message: String) {
        SVProgressHUD.showError(withStatus: message)
    }
    
    func chatManager(_ chatManager: ChatManager, didUpdateStorage message: String) {
        reloadContent()
        SVProgressHUD.dismiss()
        QBChat.instance.addDelegate(self)
    }
    
    func chatManagerWillUpdateStorage(_ chatManager: ChatManager) {
        if navigationController?.topViewController == self {
            
        }
    }
}
